devicedefender/source/DeviceDefender.cpp

```diff
@@ -37,34 +39,37 @@ namespace Aws
             OnTaskCancelledHandler &&onCancelled,
             void *cancellationUserdata) noexcept
             : OnTaskCancelled(std::move(onCancelled)), cancellationUserdata(cancellationUserdata),
-              m_allocator(allocator), m_status(ReportTaskStatus::Ready),
-              m_taskConfig{mqttConnection.get()->GetUnderlyingConnection(),
-                           ByteCursorFromString(thingName),
-                           aws_event_loop_group_get_next_loop(eventLoopGroup.GetUnderlyingHandle()),
-                           reportFormat,
-                           aws_timestamp_convert(taskPeriodSeconds, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_NANOS, NULL),
-                           aws_timestamp_convert(
-                               networkConnectionSamplePeriodSeconds,
-                               AWS_TIMESTAMP_SECS,
-                               AWS_TIMESTAMP_NANOS,
-                               NULL),
-                           ReportTask::s_onDefenderV1TaskCancelled,
-                           this},
-              m_lastError(0)
+              m_allocator(allocator), m_status(ReportTaskStatus::Ready), m_taskConfig{nullptr}, m_owningTask{nullptr},
+              m_lastError(0), m_mqttConnection{mqttConnection}, m_eventLoopGroup(eventLoopGroup)
         {
+            (void)networkConnectionSamplePeriodSeconds;
+            struct aws_byte_cursor thingNameCursor = Crt::ByteCursorFromString(thingName);
+            m_lastError =
+                aws_iotdevice_defender_config_create(&m_taskConfig, allocator, &thingNameCursor, reportFormat);
+            if (AWS_OP_SUCCESS == m_lastError)
+            {
+                aws_iotdevice_defender_config_set_task_cancelation_fn(m_taskConfig, s_onDefenderV1TaskCancelled);
+                aws_iotdevice_defender_config_set_callback_userdata(m_taskConfig, this);
+                aws_iotdevice_defender_config_set_task_period_ns(
+                    m_taskConfig,
+                    aws_timestamp_convert(taskPeriodSeconds, AWS_TIMESTAMP_SECS, AWS_TIMESTAMP_NANOS, NULL));
+            }
+            else
+            {
+                m_lastError = aws_last_error();
+            }
         }
 
         ReportTask::ReportTask(ReportTask &&toMove) noexcept
             : OnTaskCancelled(std::move(toMove.OnTaskCancelled)), cancellationUserdata(toMove.cancellationUserdata),
               m_allocator(toMove.m_allocator), m_status(toMove.m_status), m_taskConfig(std::move(toMove.m_taskConfig)),
-              m_owningTask(toMove.m_owningTask), m_lastError(toMove.m_lastError)
+              m_owningTask(toMove.m_owningTask), m_lastError(toMove.m_lastError),
```

Comment:
Is this constructor necessary?

Reply:
I'm not sure. It was already there so I made adjustments as necessary rather than figure out if it was safe to delete.